### PR TITLE
Bump dependencies for nokogiri, ruby, and rubyzip.

### PR DESCRIPTION
**Pull request by inopinatus**
nokogiri <1.10.4 and rubyzip < 1.2.3 had CVEs and/or regressions, and Ruby 2.4.7 is the minimum maintenance level.

### CODE REVIEW

**Comment by noniq** [comment]
I’m not sure if we should drop older Ruby versions as long as the tests pass successfully. @straydogstudio @simi What do you think?

**Comment by straydogstudio** [comment]
@simi @noniq I think it depends on what we want to support. It can result in failed tests that cause confusion down the road.

**Comment by noniq** [comment]
I have a slight preference for a pragmatic approach: 

Right now, the tests still pass even on older versions of Ruby. So let’s keep this versions in the build matrix for now. If there is a pull request causing tests to fail on older versions: Check if it is trivial to adapt the PR so that the code still works on these version. If so, adapt the PR. If not, officially declare the affected versions as unsupported and delete them from the build matrix.

But as said, this is a slight preference only. Let’s wait for others to comment.

**Comment by straydogstudio** [comment]
I am fine with the pragmatic approach. It could break a lot of installs to restrict older ruby versions, so we may as well wait until it is necessary. 

Note I had to limit Travis CI for axlsx_rails to Ruby 2.4.7 or newer to get tests to pass for Rails 6. I could adjust the matrix to exclude Rails 6 from previous Ruby versions and get those other tests passing again to match. I will look at that. The gemspec requirements are not so restrictive.

**Comment by inopinatus** [comment]
PR bumped for [latest rubyzip CVE](https://github.com/rubyzip/rubyzip/pull/403).

Note that rubyzip 2.0 has a constraint of `>= 2.4` in the gemspec, and the upcoming nokogiri 1.11 milestone has their, er, equivalent. Given which I suggest there's little point maintaining test configurations for Ruby below 2.4.

For consensus I relaxed the Ruby version constraint in the PR back to `~> 2`. Not so much for security now, but I still think it's reasonable to signal to contributors that it's okay to use Ruby 2.x idioms.

**Comment by straydogstudio** [comment]
@inopinatus @noniq I think we have a few issues that have surfaced here. 

First, should tests continue to be used if they support outdated dependencies, e.g. ruby 2.2. I don't think tests should be removed provided they are allowed in the context of runtime requirements. AS @noniq described, it lets us know when a version should be officially unsupported. 

Second, should caxlsx require users to use up to date dependencies. This has consequences for the end user. Nokogiri, for example, is used all over the place. I can imagine (without any real example) of a situation where the latest caxlsx version would help even when an older unsupported nokogiri is required. Should the caxlsx gem require new versions? Or, should it deprecate them and spit out warnings? As long as the function of caxlsx is not hindered by older dependencies I don't think they should be restricted. In my experience real situations have extended insecure configurations on the way to a secure one. 

I am not, however, trying to push this as an answer. It does make things flexible for the end user while making them aware of their old dependencies. 

@inopinatus Does this make sense to you? Can you provide an argument for your pull request?

**Comment by inopinatus** [comment]
@straydogstudio I think that deliberately deciding to allow a third party to continue with an insecure package amounts to negligence and is possibly actionable in civil or criminal proceedings. To avoid liability I'd never do so in any product I work on professionally. A gem is a product that installs dependencies as part of that product, possibly to customers unaware they are doing so. They are, to all intents and purposes, distributed with it.

If someone installs my product, and it pulls in an insecure dependency, then the only way I can see to discharge my duty of care to them is to ensure that an upgrade is issued that resolves the vulnerability i.e. upgrades the dependency.

Fun fact: software licenses do not protect against being sued or prosecuted. The warranty disclaimer is a possible defense in civil proceedings, but it doesn't stop those civil proceedings occurring. So you hire lawyers and go to court, which is an horrific, expensive, and embarrassing process even when sanity and licenses and your Professional Indemnity Insurance Cover are on your side. And software licenses are worthless against charges of criminal negligence, so let's hope no-one got hurt or national secrets exfiltrated.

I don't actually think it's very likely to get sued or prosecuted over an old version of a spreadsheet generator quietly installing an insecure Nokogiri and then failing to upgrade it in later updates, but stupider legal actions have occurred, particularly in the software industry, and after ~~years~~decades of designing and shipping products, observing every dot of legal advice scrupulously is an ingrained habit for me.

The other product upside of this legal paranoia is that you can use recent features with confidence, should they prove useful.

TL;DR: if you're publishing software, depend only on maintained products, and drop everything else like a hot potato.

**Comment by inopinatus** [comment]
By revealing my position I have infected you with a memetic virus, for which my apologies.

**Comment by straydogstudio** [comment]
Ha! Yes. I'm afraid I am vulnerable, but I won't be shaving my beard. I've been in legal proceedings and they are not nice. 

If we can tag older versions, and make a 2.0.2 release to provide a stop gap, we can go forward with this pull request. I would like to provide usable upgrade paths for users of the existing gem.

**Comment by noniq** [comment]
I’ll create a PR for a release of 2.0.2

**Comment by straydogstudio** [comment]
Thanks @noniq

**Comment by inopinatus** [comment]
Note that my argument does not "hold water" for Ruby versioning, for two reasons:

* A user had to explicitly choose to install Ruby first, to install this gem. Or, Ruby was distributed to them by another party e.g. Ubuntu. Either way it's not part of the distribution of this gem.

* There's no way to do so correctly. Setting `s.required_ruby_version = '>= 2.4.8'` doesn't upgrade a vulnerable 2.5.6 to 2.5.7, and `s.required_ruby_version = '>= 2.6.5'` means the upgrade can't be installed by 2.5.7 users.

As a result I think there is no duty of care, ethical or legal, regarding Ruby version.

My view on `required_ruby_version` is that it has more to do with signalling to contributors about which syntax and standard library methods can be used, and the minimum level for testing.  Once Nokogiri and Rubyzip update, all tests on 2.3 and below will start failing, hence my suggestion of `~> 2.3` for now and probably `~> 2.4` in future.

NB: I know I write quite forcefully, but it is still only one man's opinion about how things should be done.

**Comment by straydogstudio** [comment]
@inopinatus I would far rather have a clear strong statement than a fuzzy one! No worries on my end.

**Comment by straydogstudio** [comment]
@noniq  I have created a 2.0.2 branch from the commit just before the changes we started.

**Comment by straydogstudio** [comment]
@noniq Is it sufficient as is? Or does it need changes?
